blocks/log.py

```diff
@@ -0,0 +1,241 @@
+            attributes contain the information regarding the status of
+            the training process.
+
+        """
+        pass
+
+    @property
+    def status(self):
+        """A convenient way to access the status."""
+        return self.get_status()
+
+    @abstractmethod
+    def set_default_value(self, key, value):
+        """Sets a default value for 'key'."""
+        pass
+
+    @abstractmethod
+    def get_default_value(self, key):
+        """Returns the default value set for the 'key'.
+
+        Returns
+        -------
+        The default value for the `key`, or ``None`` if not set.
+
+        """
+        pass
+
+    def add_record(self, time, key, value):
+        """Adds a record to the log.
+
+        If `value` equals to the default value for the `key`, nothing is
+        done.
+
+        """
+        self._check_time(time)
+        default_value = self.get_default_value(key)
+        if value != default_value:
+            self._add_record(time, key, value)
+
+    @abstractmethod
+    def _add_record(self, time, key, value):
+        """Adds a record to the log.
+
+        The implementation method to be overriden.
+
+        """
+        pass
+
+    def fetch_record(self, time, key):
+        """Fetches a record from the log.
+
+        If no such 'key' for the `time` is found or if the value for the
+        key is ``None``, the default value for the 'key' is returned.
+
+        """
+        self._check_time(time)
+        value = self._fetch_record(time, key)
+        if value is not None:
+            return value
+        return self.get_default_value(key)
+
+    @abstractmethod
+    def _fetch_record(self, time, key):
+        """Fetches a record from the log.
+
+        The implementation method to be overriden.
+
+        """
+        pass
+
+    def __getitem__(self, time):
+        self._check_time(time)
+        return TrainingLogRow(self, time)
+
+    @property
+    def current_row(self):
+        return self[self.status.iterations_done]
+
+    @property
+    def previous_row(self):
+        return self[self.status.iterations_done - 1]
+
+    @property
+    def last_epoch_row(self):
+        return self[self.status._epoch_ends[-1]]
+
+    @abstractmethod
+    def __iter__(self):
+        """Returns an iterator over time-key-value triples of the log."""
+        pass
+
+    @abstractmethod
+    def get_row_iterator(self, time):
+        """Returns an iterator over key-value pairs of a row of the log."""
+        pass
+
+    def _check_time(self, time):
+        if not isinstance(time, int) or time < 0:
+            raise ValueError("time must be a positive integer")
+
+
+class RAMStatus(AbstractTrainingStatus):
+    """A simple training status."""
+    def __iter__(self):
+        for attr, value in self.__dict__.items():
+            if not attr.startswith("__"):
+                yield attr, value
+
+
+class RAMTrainingLog(AbstractTrainingLog):
```

Comment:
I am also not a big fan of RAM. Let's call them `TrainingStatus` and `TrainingLog`, although I think that this position should be in future occupied by a more mature implementation.